Make native builds possible

nbuchwitz opened the pull request:
With the current approach native builds are not possible, because CROSS_COMPILE cannot be unset. ~~Fix this by only set the default compiler, if CROSS_COMPILE is not set (even not if it's empty).~~ Fix this by comparing the host's architecture with the desired one. While fixing this, also some tabs have been substituted with spaces (for consistency).


With this merged, we will be one step further on our CI roadmap and can build our future kernels directly on target architecture (arm/arm64)

nbuchwitz committed:
Make native builds possible

With the current approach native builds are not possible, because
CROSS_COMPILE cannot be unset. Fix this by comparing the host's
architecture with the desired one. While fixing this, also some tabs
have been substituted with spaces (for consistency).

Signed-off-by: Nicolai Buchwitz <n.buchwitz@kunbus.com>

l1k commented:
Can't we just auto-detect the build host's arch with `uname` and set CROSS_COMPILE if it's not equivalent to the target arch?

We never used native builds because it was dog slow on the CM3, let alone CM1. Come to think of it though, building inside a mounted image on x86 (`chroot /path/to/mounted/loop/device /bin/sh` with qemu-arm-static installed) would probably perform halfway decently.

nbuchwitz commented:
@l1k even better and simpler from a users perspective. I've updated the PR and while at it also fixed some tabs vs spaces in this particular area of the script.

iluminat23 reviewed (changes requested):
Please don't mix white space changes with functional changes. And if you do s/\t/    / do it all over the file not partial.

nbuchwitz committed:
Fix formating

The usage of spaces vs tabs, the indentation and overall formatting is a
mess. Fix this by formatting the code with shfmt [1], which is a broadly
adapted tool for this (vscode, cli, vim, ...).

[1] https://github.com/mvdan/sh

Signed-off-by: Nicolai Buchwitz <n.buchwitz@kunbus.com>

iluminat23 commented on debian/update.sh:
```diff
@@ -73,6 +73,7 @@ export KBUILD_BUILD_USER="support"
 export KBUILD_BUILD_HOST="kunbus.com"
 export ARCH
 export CROSS_COMPILE
+# shellcheck disable=SC2016
```
Please add `Expressions don't expand in single quotes, use double quotes for that.` to the commit message

iluminat23 commented on debian/update.sh:
```diff
@@ -45,22 +45,27 @@ if [ -n "$PIKERNELMODDIR" ] ; then
+HOST_ARCH="$(uname -m)"
 ARCH=${ARCH:-arm}
 case "$ARCH" in
-        arm)
```
Why is the indention changing here?

nbuchwitz replied:
Even if I don't like this pattern (though it's one of the two most used ones): Code format tool (shfmt). From the point of view of reproducibility and consistency, it's a bullet I'm willing to take.

iluminat23 commented on debian/update.sh:
```diff
@@ -1,31 +1,31 @@
+        rsync -aHAX \
+                --files-from=<(cd linux; find . -name Makefile\* -o -name Kconfig\* -o -name \*.pl | grep -E -v '^\./debian') linux/ "$destdir/"
+        rsync -aHAX \
+                --files-from=<(cd linux; find "arch/${ARCH}"/include include scripts -type f) linux/ "$destdir/"
```
```suggestion
                --files-from=<(cd linux; find "arch/${ARCH}/include" include scripts -type f) linux/ "$destdir/"
```

nbuchwitz committed:
Prevent globbing with quotes (SC2086)

Signed-off-by: Nicolai Buchwitz <n.buchwitz@kunbus.com>

nbuchwitz committed:
Add exception for SC2016

Expressions don't expand in single quotes, use double quotes for that.

Signed-off-by: Nicolai Buchwitz <n.buchwitz@kunbus.com>

iluminat23 commented on debian/update.sh:
```diff
@@ -60,15 +60,20 @@ if [ -n "$PIKERNELMODDIR" ]; then
 case "$ARCH" in
 arm)
-    CROSS_COMPILE=${CROSS_COMPILE:-arm-linux-gnueabihf-}
+    if ! grep -q -E "armv(6l|7|7l)" <<<"$HOST_ARCH"; then
```
armv6l?

nbuchwitz replied:
Yes, made me also wonder, but @naruxde validated this for me on a Core

iluminat23 replied:
WTF

iluminat23 commented on debian/update.sh:
```diff
@@ -60,15 +60,20 @@ if [ -n "$PIKERNELMODDIR" ]; then
     ;;
 arm64)
-    CROSS_COMPILE=${CROSS_COMPILE:-aarch64-linux-gnu-}
+    if ! grep -q -E "(armv8|aarch64)" <<<"$HOST_ARCH"; then
```
Is output of uname not consistent or why do we need to match armv8 and aarch64?

nbuchwitz replied:
Yes, pure joy of possible architecture mixups: 

> On a 64-bit processor, you'd see a string starting with armv8 (or above) if the uname process itself is a 32-bit process, or aarch64 if it's a 64-bit process. 

Source: https://unix.stackexchange.com/a/136519

AFAIK only relevant if you run a 64-bit kernel on a 32-bit system

iluminat23 replied:
Yeah, but this would result in a 32-bit userland and thouhs bild a 32bit kernel ...

nbuchwitz replied:
Right. Fixed this and now it's only `aarch64`
